src/Components/WebAssembly/WebAssembly/src/Infrastructure/WebAssemblyJSDataStream.cs

```diff
@@ -0,0 +1,89 @@
+    {
+        private readonly DefaultWebAssemblyJSRuntime _webAssemblyJSRuntime;
+        private readonly SemaphoreSlim receiveDataSemaphore = new(initialCount: 1, maxCount: 1);
+
+        internal static async ValueTask<WebAssemblyJSDataStream> CreateWebAssemblyJSDataStreamAsync(
+            DefaultWebAssemblyJSRuntime runtime,
+            IJSStreamReference jsStreamReference,
+            long totalLength,
+            long chunkSize,
+            TimeSpan defaultCallTimeout,
+            long pauseIncomingBytesThreshold = -1,
+            long resumeIncomingBytesThreshold = -1,
+            CancellationToken cancellationToken = default)
+        {
+            var streamId = runtime.JSDataStreamNextInstanceId++;
+            var webAssemblyJSDataStream = new WebAssemblyJSDataStream(runtime, streamId, totalLength, defaultCallTimeout, pauseIncomingBytesThreshold, resumeIncomingBytesThreshold, cancellationToken);
+            var dotnetStreamReference = DotNetObjectReference.Create(webAssemblyJSDataStream);
+            await runtime.InvokeVoidAsync("Blazor._internal.sendJSDataStreamUsingObjectReference", jsStreamReference, streamId, chunkSize, dotnetStreamReference);
+            return webAssemblyJSDataStream;
+        }
+
+        private WebAssemblyJSDataStream(
+            DefaultWebAssemblyJSRuntime runtime,
+            long streamId,
+            long totalLength,
+            TimeSpan jsInteropDefaultCallTimeout,
+            long pauseIncomingBytesThreshold,
+            long resumeIncomingBytesThreshold,
+            CancellationToken cancellationToken) :
+            base(
+                runtime.JSDataStreamInstances,
+                streamId,
+                totalLength,
+                jsInteropDefaultCallTimeout,
+                pauseIncomingBytesThreshold,
+                resumeIncomingBytesThreshold,
+                cancellationToken)
+        {
+            _webAssemblyJSRuntime = runtime;
+        }
+
+        protected override void RaiseUnhandledException(Exception exception)
+        {
+            throw exception;
+        }
+
+        [JSInvokable("ReceiveJSDataChunk")]
+        public async Task<bool> ReceiveJSDataChunk(long streamId, long chunkId, string error)
+        {
+            // Ensure that the DotNetDataReference still points to an active stream
+            if (!_webAssemblyJSRuntime.JSDataStreamInstances.ContainsKey(streamId))
+            {
+                // There is no data stream with the given identifier. It may have already been disposed.
+                // We notify JS that the stream has been cancelled/disposed.
+                return false;
+            }
+
+            var data = Array.Empty<byte>();
+            if (string.IsNullOrEmpty(error))
+            {
+                // Ideally this byte array would be transferred directly as a parameter on this
+                // call, however that's not currently possible due to: https://github.com/dotnet/runtime/issues/53378
```

Comment:
I'm confused about this. I thought that https://github.com/dotnet/runtime/issues/53378 referred to an underlying issue in the runtime which, for `[JSInvokable]`, we've already implemented our own workaround. If that is the case, and since this is a `[JSInvokable]` method, why doesn't it work it receive a `byte[]` as a regular parameter here?

If this is the case, can we simplify things by unifying the implementation between webassembly and webview, since they both should just be able to build on the binary interop support?

Or am I missing something?